grammar.js

```diff
@@ -193,35 +193,40 @@ module.exports = grammar({
     assert_msg: ($) => seq(",", $.expr),
 
     type: ($) =>
+      choice(
+        $.simple_type,
+        "timer",
+        seq("union", "{", list1($.type, ","), "}"),
+        seq("function", $.func_params),
+        seq("event", "(", optional($.formal_args), ")"),
+        seq("hook", "(", optional($.formal_args), ")"),
+        $.id,
```

Comment:
The important trick here is to have `$.id` and `function`/`event`/`hook` on this level, not below. IMO rules like `timer` or the one for union don't really belong here, so I'd suggest to move them down into the following rule.